synapse/handlers/directory.py

```diff
@@ -141,10 +141,15 @@ async def create_association(
                         403, "You must be in the room to create an alias for it"
                     )
 
-            if not await self.spam_checker.user_may_create_room_alias(
+            spam_check = await self.spam_checker.user_may_create_room_alias(
                 user_id, room_alias
-            ):
-                raise AuthError(403, "This user is not permitted to create this alias")
+            )
+            if isinstance(spam_check, Codes):
+                raise AuthError(
+                    403,
+                    "This user is not permitted to publish rooms to the room list",
+                    spam_check,
+                )
```

Comment:
I think something like this:

```suggestion
            if spam_check != self.spam_checker.NOT_SPAM:
                if isinstance(spam_check, Codes):
                    errcode = spam_check
                else:
                    errcode = Codes.FORBIDDEN
                
                raise AuthError(
                    403,
                    "This user is not permitted to publish rooms to the room list",
                    spam_check,
                )
```

despite being a bit more of a mouthful, is a bit clearer on how we expect spam checks to fail and makes it less likely that we introduce bugs in the future by ignoring potential future new return values for failures.

Also note that the error message here is not right (we're checking if we can create an alias, not if we can publish to the room directory).

Reply:
So, we're normalizing the return value twice? Once in spamcheck.py and one in the caller code? Is that expected?

Reply:
mypy isn't happy with that

```
> synapse/handlers/directory.py:151: error: Statement is unreachable  [unreachable]
```